### PR TITLE
[BEAM-10671] Add environment configuration fields as first-class pipeline options.

### PR DESCRIPTION
**Pull request by ibzib**
This PR only changes the Python SDK. The Java and Go SDKs also have
these options, which can be changed in future PRs.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
Build python source distribution and wheels

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by ibzib** [comment]
Run Python PreCommit

**Comment by ibzib** [comment]
Note that I didn't add `os` or `architecture` for process mode, since they don't seem to do anything. Please correct me if I'm wrong about that.

R: @robertwb @mxm

**Comment by ibzib** [comment]
Python precommit is failing but not terminating (BEAM-10751).

I'm addressing the cause of the failures with an improved validator implementation.

**Comment by ibzib** [comment]
retest this please

**Comment by ibzib** [comment]
Run Spotless PreCommit

**Comment by ibzib** [comment]
Run Python PreCommit

**Comment by ibzib** [comment]
Run PythonDocker PreCommit

**Comment by ibzib** [comment]
Run Portable_Python PreCommit

**Comment by ibzib** [comment]
Run Java PreCommit

**Comment by ibzib** [comment]
Run PythonDocker PreCommit

**Comment by ibzib** [comment]
Run Portable_Python PreCommit

**Comment by ibzib** [comment]
Python2_PVR_Flink and Website precommits have passed, but the display here is stale.

R: @angoenka

**Comment by ibzib** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by ibzib** [comment]
Run Website PreCommit

**Comment by ibzib** [comment]
ping @robertwb @mxm @angoenka 

Can someone please review this?

File: sdks/python/apache_beam/options/pipeline_options.py
Comment (mxm) [inline sdks/python/apache_beam/options/pipeline_options.py:1052]: Could we change the option name pattern to `env_<type>_<option>`?

Comment (ibzib) [reply sdks/python/apache_beam/options/pipeline_options.py]: Done.

**Comment by mxm** [comment]
Thanks! I'll have another look tomorrow but generally looks good to me.

**Comment by ibzib** [comment]
(Filed BEAM-10901 for test flakes.)

**Review by mxm — COMMENTED** [review]
I'm a little unsure about the overhead this adds in terms of the number of options. Would it make sense to instead use the following format? For example:

```
--environment_type PROCESS
--environment_option command=cmd
--environment_option variable='MY_ENV=XY'
```

This would make it more clear what options belong to the environment. The current state of the PR would look like the following:

```
--environment_type PROCESS
--env_process_command cmd
--env_process_variable 'MY_ENV=XY'
```

My concern is that it is confusing to see the relationship between the environment type and the options.

**Comment by ibzib** [comment]
> I'm a little unsure about the overhead this adds in terms of the number of options. Would it make sense to instead use the following format? For example:
> 
> ```
> --environment_type PROCESS
> --environment_option command=cmd
> --environment_option variable='MY_ENV=XY'
> ```
> 
> This would make it more clear what options belong to the environment. The current state of the PR would look like the following:
> 
> ```
> --environment_type PROCESS
> --env_process_command cmd
> --env_process_variable 'MY_ENV=XY'
> ```
> 
> My concern is that it is confusing to see the relationship between the environment type and the options.

I like how this way creates a clearer hierarchy of options. I drafted an implementation based on how we handle --experiments. If you approve of this method, I will finish up the documentation work later.

File: sdks/python/apache_beam/options/pipeline_options_validator_test.py
Comment (mxm) [inline sdks/python/apache_beam/options/pipeline_options_validator_test.py:629]: We could think about dropping the `env_` prefix for the options, e.g. `container_image` instead of `docker_container_image`.

**Review by mxm — COMMENTED** [review]
Thanks @ibzib! I like the new approach better. It makes it explicit which options have been defined for an environment.

File: sdks/python/apache_beam/options/pipeline_options_validator_test.py
Comment (ibzib) [reply sdks/python/apache_beam/options/pipeline_options_validator_test.py:629]: I want to keep the prefix as a hint for users. This way it is immediately obvious which environment type each option corresponds to.

**Comment by ibzib** [comment]
Run PortableJar_Flink PostCommit

**Comment by ibzib** [comment]
More flakes: BEAM-10866

File: sdks/python/apache_beam/options/pipeline_options_validator_test.py
Comment (mxm) [reply sdks/python/apache_beam/options/pipeline_options_validator_test.py:629]: Sounds fair.

**Review by mxm — COMMENTED** [review]
This needs a rebase and a documentation update. Otherwise LGTM.

**Comment by ibzib** [comment]
Run Python PreCommit

**Comment by ibzib** [comment]
> This needs a rebase and a documentation update. Otherwise LGTM.

@mxm done

**Review by mxm — COMMENTED** [review]
Thanks @ibzib. I'm not sure we should merge this because this would result in a different behavior for Python. Could we also make sure this works the same way in Java and Go? 

If you think this should be tackled in a follow-up, then I suggest to revert the documentation change, or add a note that there are two ways to configure the environment.

**Comment by ibzib** [comment]
> If you think this should be tackled in a follow-up, then I suggest to revert the documentation change, or add a note that there are two ways to configure the environment.

@mxm I want to get this merged to unblock #12385. I've reverted the documentation changes for now.

**Review by mxm — APPROVED** [review]
LGTM. Will merge if tests pass.

**Comment by ibzib** [comment]
Run Portable_Python PreCommit

**Comment by ibzib** [comment]
Run Python PreCommit

**Comment by ibzib** [comment]
Run Python PreCommit

**Comment by ibzib** [comment]
Run Python PreCommit

**Comment by ibzib** [comment]
Python precommit failures seem to be unrelated: BEAM-10985

I won't merge until it passes, since it seems the cause of failure is that no unit tests are running.

**Comment by mxm** [comment]
Run Portable_Python PreCommit

**Comment by mxm** [comment]
Run Python PreCommit

**Pull request merged**
